File: scripts/generate-flocken-imagen-seeds.js
/**
 * Generate 10 deterministic images on Vertex AI using Imagen + seeds.
 * Node.js-version – kör med: node scripts/generate-flocken-imagen-seeds.js
 *
 * Prereqs:
 * - npm install @google/genai
 * - gcloud auth application-default login (eller GOOGLE_APPLICATION_CREDENTIALS)
 */

const fs = require('fs');
const path = require('path');
const { GoogleGenAI } = require('@google/genai');
const dotenv = require('dotenv');

// Load .env.local if present (for VERTEX_GEMINI_API_KEY)
dotenv.config({ path: path.join(__dirname, '..', '.env.local') });

// ========= CONFIG =========
const PROJECT_ID = process.env.GCP_PROJECT || 'nastahem-tracking';
const LOCATION = process.env.GCP_LOCATION || 'europe-west4';
const MODEL_NAME = 'imagen-4.0-generate-001';

const PROMPT = `
Photorealistic advertising image in bright, warm daylight in a Swedish everyday park.
A dog owner (age 25–40) stands casually holding a smartphone; calm, friendly mood (pleasant and upbeat, not exaggerated).
A friendly dog (labrador or poodle) sits next to them looking up.
Modern smartphone photography look: sharp focus on dog and hand with phone, soft background bokeh, clean colors, natural lighting, no harsh shadows.
No recognizable landmarks. No text. No logos.
Phone screen must not be readable; only a vague hint of a map/app.
Composition with generous negative space for future overlay.
`.trim();

const NEGATIVE_PROMPT = `
dystopian, gloomy, rainy, night, harsh shadows, serious faces, exaggerated laughter,
posed stock photo, extra fingers, deformed hands, text, watermark, readable UI, logo, brand marks
`.trim();

const SEEDS = [101, 202, 303, 404, 505, 606, 707, 808, 909, 1001];

// ========= RUN =========
async function main() {
  const scriptDir = __dirname;
  const outDir = path.join(scriptDir, '..', 'generated_images');

  // Autentisering - Tvinga Vertex AI (Project + Service Account)
  let keyPath = null;
  if (!process.env.GOOGLE_APPLICATION_CREDENTIALS) {
    // Försök hitta email-signup-api-service key först (har Vertex AI User-roll)
    const emailSignupKey = path.join(scriptDir, 'email-signup-api-service-key.json');
    if (fs.existsSync(emailSignupKey)) {
      keyPath = emailSignupKey;
    } else {
      // Fallback: försök hitta alla key-filer och kolla client_email
      const keyFiles = fs.readdirSync(scriptDir)
        .filter(f => f.endsWith('-key.json') || f.endsWith('-credentials.json'))
        .map(f => path.join(scriptDir, f));

      for (const kf of keyFiles) {
        try {
          const keyData = JSON.parse(fs.readFileSync(kf, 'utf8'));
          if (keyData.client_email && keyData.client_email.includes('email-signup-api-service')) {
            keyPath = kf;
            break;
          }
        } catch (e) {
          // Ignorera ogiltiga JSON-filer
        }
      }

      // Om ingen email-signup hittades, använd nastahem-tracking-key.json som fallback
      if (!keyPath) {
        const fallbackKey = path.join(scriptDir, 'nastahem-tracking-key.json');
        if (fs.existsSync(fallbackKey)) {
          keyPath = fallbackKey;
        }
      }
    }

    if (keyPath) {
      process.env.GOOGLE_APPLICATION_CREDENTIALS = keyPath;
      console.log('🔑 Använder service account:', path.basename(keyPath));
    } else {
      console.log('⚠️  Ingen service account key hittad – använder application-default credentials');
    }
  }

  fs.mkdirSync(outDir, { recursive: true });

  console.log('\n🔧 Initierar Vertex AI...');
  console.log('   Projekt:', PROJECT_ID);
  console.log('   Region:', LOCATION);
  console.log('   Model:', MODEL_NAME);
  console.log('   Output:', outDir);
  console.log('   Watermark: ❌ Avstängt (för determinism)\n');

  const ai = new GoogleGenAI({
    vertexai: true,
    project: PROJECT_ID,
    location: LOCATION,
  });

  console.log('🔐 Auth: Vertex AI (ADC/service account)');

  const stamp = new Date().toISOString().replace(/[-:]/g, '').slice(0, 15);

  console.log(`🎨 Genererar ${SEEDS.length} bilder med seeds...\n`);

  for (let i = 0; i < SEEDS.length; i++) {
    const seed = SEEDS[i];
    process.stdout.write(`[${i + 1}/${SEEDS.length}] Genererar med seed ${seed}... `);

    try {
      const config = {
        numberOfImages: 1,
        aspectRatio: '1:1',
        seed,
        addWatermark: false,
        negativePrompt: NEGATIVE_PROMPT,
      };

      const response = await ai.models.generateImages({
        model: MODEL_NAME,
        prompt: PROMPT,
        config,
      });

      const img = response?.generatedImages?.[0]?.image;
      if (!img?.imageBytes) {
        throw new Error('Ingen bild i svaret: ' + JSON.stringify(response?.generatedImages?.[0]));
      }

      const filename = `flocken_v1_seed${seed}_${stamp}.png`;
      const filepath = path.join(outDir, filename);
      const buffer = Buffer.from(img.imageBytes, 'base64');
      fs.writeFileSync(filepath, buffer);

      console.log('✅ Sparad:', filename);
    } catch (err) {
      const msg = err.message || String(err);
      console.log('❌ Fel:', msg.slice(0, 120));
      if (msg.includes('PERMISSION_DENIED') || msg.includes('403')) {
        console.log('   → Ge service account rollen "Vertex AI User" i GCP, eller kör: gcloud auth application-default login');
      }
      if (process.env.DEBUG) console.error(err);
    }
  }

  console.log('\n✅ Klart! Bilder i:', outDir);
}

main().catch((err) => {
  console.error('\n❌ OVÄNTAT FEL:', err);
  process.exit(1);
});

File: scripts/generate-flocken-imagen-seeds.py
"""
Generate 10 deterministic images on Vertex AI using Imagen + seeds.

Prereqs:
- pip install --upgrade google-cloud-aiplatform
- gcloud auth application-default login
- Set PROJECT_ID + LOCATION
"""

import os
import sys
import traceback
from datetime import datetime

import vertexai

# Stöd för både preview och GA vision_models (SDK-versioner skiljer)
try:
    from vertexai.preview.vision_models import ImageGenerationModel
except ImportError:
    from vertexai.vision_models import ImageGenerationModel

# ========= CONFIG =========
PROJECT_ID = os.getenv("GCP_PROJECT", "nastahem-tracking")  # Uppdaterat till Flocken-projektet
LOCATION = os.getenv("GCP_LOCATION", "europe-west4")  # pick your Vertex region
MODEL_NAME = "imagen-4.0-generate-001"  # see supported model versions in docs
OUT_DIR = os.getenv("OUT_DIR", "./generated_images")

# Master prompt (same idea as Nano Banana prompt, but tuned for Imagen)
PROMPT = """
Photorealistic advertising image in bright, warm daylight in a Swedish everyday park.
A dog owner (age 25–40) stands casually holding a smartphone; calm, friendly mood (pleasant and upbeat, not exaggerated).
A friendly dog (labrador or poodle) sits next to them looking up.
Modern smartphone photography look: sharp focus on dog and hand with phone, soft background bokeh, clean colors, natural lighting, no harsh shadows.
No recognizable landmarks. No text. No logos.
Phone screen must not be readable; only a vague hint of a map/app.
Composition with generous negative space for future overlay.
""".strip()

NEGATIVE_PROMPT = """
dystopian, gloomy, rainy, night, harsh shadows, serious faces, exaggerated laughter,
posed stock photo, extra fingers, deformed hands, text, watermark, readable UI, logo, brand marks
""".strip()

# 10 seeds (byt gärna till egna om du vill)
SEEDS = [101, 202, 303, 404, 505, 606, 707, 808, 909, 1001]

# ========= RUN =========
def main():
    # Skapa output-mapp relativt till script-mappen
    script_dir = os.path.dirname(os.path.abspath(__file__))
    out_path = os.path.join(script_dir, "..", "generated_images")
    os.makedirs(out_path, exist_ok=True)

    # Initiera Vertex AI
    # Kolla efter service account key
    key_path = os.path.join(script_dir, "nastahem-tracking-key.json")
    if os.path.exists(key_path) and not os.environ.get("GOOGLE_APPLICATION_CREDENTIALS"):
        os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = key_path
        print(f"🔑 Använder service account: {os.path.basename(key_path)}")

    print(f"🔧 Initierar Vertex AI...")
    print(f"   Projekt: {PROJECT_ID}")
    print(f"   Region: {LOCATION}")
    print(f"   Model: {MODEL_NAME}")
    print(f"   Output: {out_path}")
    print(f"   Watermark: ❌ Avstängt (för determinism)\n")

    vertexai.init(project=PROJECT_ID, location=LOCATION)
    model = ImageGenerationModel.from_pretrained(MODEL_NAME)

    stamp = datetime.now().strftime("%Y%m%d_%H%M%S")

    print(f"🎨 Genererar {len(SEEDS)} bilder med seeds...\n")

    for i, seed in enumerate(SEEDS, 1):
        print(f"[{i}/{len(SEEDS)}] Genererar med seed {seed}...", end=" ")
        
        try:
            response = model.generate_images(
                prompt=PROMPT,
                negative_prompt=NEGATIVE_PROMPT,
                number_of_images=1,
                aspect_ratio="1:1",      # ändra till "4:5" eller "9:16" i separata körningar
                seed=seed,
                add_watermark=False,     # required for deterministic behavior per docs
            )

            img = response.images[0]
            filename = f"flocken_v1_seed{seed}_{stamp}.png"
            filepath = os.path.join(out_path, filename)

            # Spara: vissa SDK-versioner använder save(path), andra save(location=path)
            try:
                img.save(location=filepath)
            except TypeError:
                img.save(filepath)

            print(f"✅ Sparad: {filename}")

        except Exception as e:
            print(f"❌ Fel: {e}")
            traceback.print_exc()
            continue

    print(f"\n✅ Klart! {len(SEEDS)} bilder genererade i: {out_path}")

if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print("\n" + "=" * 60)
        print("❌ OVÄNTAT FEL – kopia hela utskriften nedan om du behöver hjälp:")
        print("=" * 60)
        traceback.print_exc()
        sys.exit(1)
